src/array/cpu/segment_reduce.h

```diff
@@ -101,6 +103,58 @@ void ScatterAdd(NDArray feat, NDArray idx, NDArray out) {
   }
 }
 
+/*!
+ * \brief CPU kernel of Scatter Add (on first dimension) operator for heterogeneous graph.
+ * \note math equation: out[idx[i], *] += feat[i, *]
+ * \param feat The input tensor.
+ * \param idx The indices tensor.
+ * \param out The output tensor.
+ */
+
+template <typename IdType, typename DType>
+void UpdateGradMinMax_hetero(HeteroGraphPtr graph,
+                       std::vector<NDArray> list_feat,
+                       std::vector<NDArray> list_idx,
+                       std::vector<NDArray> list_idx_ntypes,
+                       std::vector<NDArray> list_out) {
+  std::vector<std::vector<dgl_id_t>> dst_src_ntids(graph->NumVertexTypes(),
+    std::vector<dgl_id_t>());
+  for (dgl_type_t etype = 0; etype < graph->NumEdgeTypes(); ++etype) {
+    auto pair = graph->meta_graph()->FindEdge(etype);
+    const dgl_id_t dst_id = pair.first;  // graph is reveresed
+    const dgl_id_t src_id = pair.second;
+    dst_src_ntids[dst_id].push_back(src_id);  // can have duplicates. Use Hashtable to optimize.
+  }
+  std::vector<bool> updated(graph->NumVertexTypes(), false);
+  for (int dst_id = 0; dst_id < dst_src_ntids.size(); ++dst_id) {
+    updated[dst_id] = false;
+    for (int j = 0; j < dst_src_ntids[dst_id].size(); ++j) {
+      int src_id = dst_src_ntids[dst_id][j];
+      if (updated[src_id]) continue;
```

Comment:
So if a source type is already updated given the destination node type, the update will skip.  What if there are multiple edge types connecting between the same source/destination node type?  I don't think this is right but correct me if I'm wrong.

Reply:
We store the source node type or edge-type of the max/min value for a destination type after comparing all the edge types. So, I think we need to update only once. The current unittest covers this step:
        `('user', 'plays', 'game'): ([0, 1, 2, 1], [0, 0, 1, 1]),
        ('user', 'wishes', 'game'): ([0, 1, 1], [0, 0, 1]),`

Reply:
I see your point. I have the issue with `copy_e`. When multiple edge types share the same `src` and `dst` node type, only one etype computes the gradient instead of both. Maybe we can check the `op` to  see whether its `copy_u` or `copy_e`.

Reply:
I believe the same issue will occur for `copy_u`.  Say if you have `('user', 'clicks', 'item')` and `('user', 'likes', 'item')`, you have two edge types sharing the same source node type and the same destination node type.  In this case your code will only compute the gradient through `clicks` but not `likes` for both cases.

Reply:
Another question though is regarding to the following:
> If multiple edge types share the same src but different dst node types, the partial derivatives of all the dst type needs to be summed and update src node type. For example, in the case of ('user', 'plays', 'game') and ('user', 'follows, 'user'), node type user will get updates from both game and user. That's why we clear dst_id's flag updated[dst_id] = false; to get update from all dsttypes.

In this case, I still don't understand why you can skip `src_id` when two edge types share the same source type but different destination type.

Say that you have edge type `(user, clicks, item)` and `(user, likes, vendor)`, and both edge types have edges (0,0) and (0,0) (meaning that user 0 clicks item 0 and user 0 likes vendor 0).  Running a `copy_u_max` will give both item 0 and vendor 0 the same value from user 0.  In backward propagation, user 0's gradient will be the sum of item 0's gradient and vendor 0's gradient.  With your implementation though, `updated[user]` will be set to True after `clicks` relation is computed.  When it comes to the edge type `vendor`, the computation will skip because `updated[user]` is True.  So user 0's gradient will be only from item 0.  Correct?  Or am I missing something?

Reply:
You are correct! Fixed now.

Reply:
You are correct. Fixed!